File: OJ/leet208/leet208.cpp
/*
leet208

Trie

*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <utility>
// #include "../utils/LeetCpp.utils.hpp"

using namespace std;

struct Node
{
    bool isleaf;
    vector<Node*> next;

    Node(bool isleaf=false): isleaf(isleaf), next(26, nullptr) {}
};


class Trie {
public:
    Node* root;
    /** Initialize your data structure here. */
    Trie(): root{new Node()} {

    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        auto p = root;
        for (const auto& ch: word) {
            if (p->next[ch-'a'] == nullptr) 
                p->next[ch-'a'] = new Node();
            p = p->next[ch-'a'];
        }
        p->isleaf = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        auto p = root;
        for (const auto& ch: word) {
            if (p->next[ch-'a'] == nullptr) 
                return false;
            p = p->next[ch-'a'];
        }
        return p->isleaf;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        auto p = root;
        for (const auto& ch: prefix) {
            if (p->next[ch-'a'] == nullptr) 
                return false;
            p = p->next[ch-'a'];
        }
        return true;
    }
};


int main(int argc, char const *argv[])
{
    cout << boolalpha;

    Trie* tree = new Trie();
    tree->insert("apple");
    cout << tree->search("apple") << endl;
    cout << tree->search("app") << endl;
    cout << tree->startsWith("app") << endl;

    tree->insert("app");
    cout << tree->search("app") << endl;

    return 0;
}

File: OJ/leet208/leet208.py
# leet208

from typing import List


class Node:
    def __init__(self, isleaf=False) -> None:
        self.next = {}
        self.isleaf = isleaf

class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = Node()


    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        p = self.root
        for ch in word:
            if ch not in p.next: p.next[ch] = Node()
            p = p.next[ch]
        p.isleaf = True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        p = self.root
        for ch in word:
            if ch not in p.next: return False
            p = p.next[ch]
        return p.isleaf 


    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        p = self.root
        for ch in prefix:
            if ch not in p.next: return False
            p = p.next[ch]
        return True


trie = Trie()

print(trie.insert("apple"))
print(trie.search("apple"))
print(trie.search("app"))
print(trie.startsWith("app"))
print(trie.insert("app"))
print(trie.search("app"))



